subprojects/core/src/main/java/org/gradle/internal/service/scopes/CoreBuildSessionServices.java

```diff
@@ -158,15 +161,37 @@ BuildStartedTime createBuildStartedTime(Clock clock, BuildRequestMetaData buildR
         return BuildStartedTime.startingAt(Math.min(currentTime, buildRequestMetaData.getStartTime()));
     }
 
+    /**
+     * ClientExecHandleFactory is available in global scope, but we need to provide it here to use a build dir based file resolver.
+     */
     @Provides
-    protected ExecFactory decorateExecFactory(ExecFactory execFactory, FileResolver fileResolver, FileCollectionFactory fileCollectionFactory, Instantiator instantiator, BuildCancellationToken buildCancellationToken, ObjectFactory objectFactory, JavaModuleDetector javaModuleDetector) {
+    ClientExecHandleFactory createExecHandleFactory(
```

Comment:
❓  I think I understand what you mean here. So we have a global instance of this service, but in this scope we need a specialized variant of it, right?

Reply:
Yes, exactly. We create `ClientExecHandleFactory` in `BasicGlobalScopeServices` but there we have only `IdentityFileResolver`. But in lower scopes we want to instance with `BaseDirFileResolver`.